### PR TITLE
fix: ensure ANDed matches for ratelimit rules

### LINKED ISSUE #2564: AND Header matches in ratelimiting not working
*Description*:

Looks like a regression, unable to get ratelimiting to work when multiple header matches are ANDed together

Followed https://gateway.envoyproxy.io/latest/user/global-rate-limit/ and updated the `policy-httproute` to add 2 header matches

```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: BackendTrafficPolicy
metadata:
  name: policy-httproute
  namespace: default
spec:
  rateLimit:
    global:
      rules:
      - clientSelectors:
        - headers:
          - name: x-user-id
            type: Exact
            value: one
          - name: x-user-org
            type: Exact
            value: acme
        limit:
          requests: 3
          unit: Hour
    type: Global
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: http-ratelimit
    namespace: default
```

```
for i in {1..4}; do curl -I --header "Host: ratelimit.example" --header "x-user-id: one" --header "x-user-org: acme" http://localhost/get ; sleep 1; done
HTTP/1.1 200 OK
content-type: application/json
x-content-type-options: nosniff
date: Mon, 05 Feb 2024 17:31:18 GMT
content-length: 577
x-envoy-upstream-service-time: 0
server: envoy

HTTP/1.1 200 OK
content-type: application/json
x-content-type-options: nosniff
date: Mon, 05 Feb 2024 17:31:19 GMT
content-length: 577
x-envoy-upstream-service-time: 0
server: envoy

HTTP/1.1 200 OK
content-type: application/json
x-content-type-options: nosniff
date: Mon, 05 Feb 2024 17:31:20 GMT
content-length: 577
x-envoy-upstream-service-time: 0
server: envoy

HTTP/1.1 200 OK
content-type: application/json
x-content-type-options: nosniff
date: Mon, 05 Feb 2024 17:31:21 GMT
content-length: 577
x-envoy-upstream-service-time: 0
server: envoy
```

The last  request should have been ratelimited

### PR DESCRIPTION
**Pull request by shawnh2**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #2564

### CODE REVIEW

**Comment by shawnh2** [comment]
ci blocked by #2608

File: test/e2e/tests/ratelimit.go [hunk 1]
```diff
@@ -74,6 +75,96 @@ var RateLimitTest = suite.ConformanceTest{
...
+					Headers:    ratelimitHeader,
+				},
+				Namespace: ns,
+			}
+			expectOkResp.Response.Headers["X-Ratelimit-Limit"] = "3, 3;w=3600"
+			expectOkReq := http.MakeRequest(t, &expectOkResp, gwAddr, "HTTP", "http")
+
+			expectLimitResp := http.ExpectedResponse{
+				Request: http.Request{
+					Path:    "/",
```
Comment (arkodg) [inline test/e2e/tests/ratelimit.go:111]: shouldnt the requests be for `/get` ?

File: test/e2e/tests/ratelimit.go [hunk 2]
```diff
@@ -74,6 +75,96 @@ var RateLimitTest = suite.ConformanceTest{
...
+				Response: http.Response{
+					StatusCode: 200,
+					Headers:    ratelimitHeader,
+				},
+				Namespace: ns,
+			}
+			expectOkResp.Response.Headers["X-Ratelimit-Limit"] = "3, 3;w=3600"
+			expectOkReq := http.MakeRequest(t, &expectOkResp, gwAddr, "HTTP", "http")
+
+			// should just send exactly 4 requests, and still expect 200
```
Comment (arkodg) [inline test/e2e/tests/ratelimit.go:155]:
```suggestion
			// send exactly 3 requests, and still expect 200
```

File: test/e2e/testdata/ratelimit-header-match.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        - clientSelectors:
+            - headers:
+                - name: x-user-id
+                  type: Exact
+                  value: one
+                - name: x-user-org
+                  type: Exact
+                  value: acme
+          limit:
+            requests: 3
```
Comment (arkodg) [inline test/e2e/testdata/ratelimit-header-match.yaml:25]: are you testing for 3 requests/hour or 4 requests/hr ?

File: test/e2e/tests/ratelimit.go [hunk 4]
```diff
@@ -74,6 +75,96 @@ var RateLimitTest = suite.ConformanceTest{
...
+
+			// keep sending requests till get 200 first, that will cost one 200
+			http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, expectOkResp)
+
+			// fire the rest request
```
Comment (arkodg) [inline test/e2e/tests/ratelimit.go:160]:
```suggestion
			// fire the rest of the requests
```

File: test/e2e/tests/ratelimit.go [hunk 5]
```diff
@@ -74,6 +75,96 @@ var RateLimitTest = suite.ConformanceTest{
...
+			if err := GotExactExpectedResponse(t, 2, suite.RoundTripper, expectOkReq, expectOkResp); err != nil {
+				t.Errorf("fail to get expected response at first three request: %v", err)
+			}
+			if err := GotExactExpectedResponse(t, 1, suite.RoundTripper, expectLimitReq, expectLimitResp); err != nil {
+				t.Errorf("fail to get expected response at last fourth request: %v", err)
```
Comment (arkodg) [inline test/e2e/tests/ratelimit.go:131]:
```suggestion
				t.Errorf("failed to get expected response for the last (fourth) request: %v", err)
```

File: test/e2e/tests/ratelimit.go [hunk 6]
```diff
@@ -74,6 +75,96 @@ var RateLimitTest = suite.ConformanceTest{
...
+			expectLimitReq := http.MakeRequest(t, &expectLimitResp, gwAddr, "HTTP", "http")
+
+			// should just send exactly 4 requests, and expect 429
+
+			// keep sending requests till get 200 first, that will cost one 200
+			http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, expectOkResp)
+
+			// fire the rest request
+			if err := GotExactExpectedResponse(t, 2, suite.RoundTripper, expectOkReq, expectOkResp); err != nil {
+				t.Errorf("fail to get expected response at first three request: %v", err)
```
Comment (arkodg) [inline test/e2e/tests/ratelimit.go:128]:
```suggestion
				t.Errorf("fail to get expected response for the first three requests: %v", err)
```

**Review by arkodg — DISMISSED** [review]
LGTM thanks !

**Review by arkodg — APPROVED** [review]

**Review by AliceProxy — APPROVED** [review]

**Pull request merged**